drivers/net/netvsc/hn_ethdev.c

```diff
@@ -1129,12 +1098,146 @@ hn_detach(struct hn_data *hv)
 	hn_rndis_detach(hv);
 }
 
+static int
+hn_dev_mtu_set(struct rte_eth_dev *dev, uint16_t mtu)
+{
+	struct hn_data *hv = dev->data->dev_private;
+	struct hn_rx_queue **rxqs = (struct hn_rx_queue **)dev->data->rx_queues;
+	struct hn_tx_queue **txqs = (struct hn_tx_queue **)dev->data->tx_queues;
+	struct rte_eth_dev *vf_dev;
+	unsigned int orig_mtu = dev->data->mtu;
+	uint32_t rndis_mtu;
+	int ret = 0;
+	int i;
+
+	if (dev->data->dev_started) {
+		PMD_DRV_LOG(ERR, "Device must be stopped before changing MTU");
+		return -EIO;
+	}
+
+	/* Change MTU of underlying VF netdev first, if it exists */
+	rte_rwlock_read_lock(&hv->vf_lock);
+	vf_dev = hn_get_vf_dev(hv);
+	if (hv->vf_ctx.vf_vsc_switched && vf_dev) {
+		ret = vf_dev->dev_ops->mtu_set(vf_dev, mtu);
+		if (ret) {
+			rte_rwlock_read_unlock(&hv->vf_lock);
+			goto out;
+		}
+	}
+	rte_rwlock_read_unlock(&hv->vf_lock);
+
+	/* Release channel resources */
+	hn_detach(hv);
+
+	/* Close vmbus channels */
+	for (i = 0; i < hv->num_queues; i++)
+		rte_vmbus_chan_close(hv->channels[i]);
+
+	/* Unmap and re-map vmbus device */
+	rte_vmbus_unmap_device(hv->vmbus);
+	ret = rte_vmbus_map_device(hv->vmbus);
+	if (ret) {
+		/* This is a catastrophic error - the device is unusable */
+		PMD_DRV_LOG(ERR, "Could not re-map vmbus device!");
+		return ret;
+	}
+
+	/* Update pointers to re-mapped UIO resources */
+	hv->rxbuf_res = hv->vmbus->resource[HV_RECV_BUF_MAP];
+	hv->chim_res  = hv->vmbus->resource[HV_SEND_BUF_MAP];
+
+	/* Re-open the vmbus channel */
+	ret = rte_vmbus_chan_open(hv->vmbus, &hv->channels[0]);
+	if (ret) {
+		/* This is a catastrophic error - the device is unusable */
+		PMD_DRV_LOG(ERR, "Could not re-open vmbus channel!");
+	}
+
+	rte_vmbus_set_latency(hv->vmbus, hv->channels[0], hv->latency);
+
+	/* Point primary queues at new primary channel */
+	rxqs[0]->chan = hv->channels[0];
+	txqs[0]->chan = hv->channels[0];
+
+	ret = hn_attach(hv, mtu);
+	if (ret)
+		goto error;
+
+	/* Create vmbus subchannels, additional RNDIS configuration */
+	ret = hn_dev_configure(dev);
+	if (ret)
+		goto error;
+	
+	/* Point any additional queues at new subchannels */
+	for (i = 1; i < dev->data->nb_rx_queues; i++)
+		rxqs[i]->chan = hv->channels[i];
+	for (i = 1; i < dev->data->nb_tx_queues; i++)
+		txqs[i]->chan = hv->channels[i];
+
+out:
+	if (hn_rndis_get_mtu(hv, &rndis_mtu))
+		PMD_DRV_LOG(ERR, "Could not get MTU via RNDIS");
+	else {
+		dev->data->mtu = (uint16_t)rndis_mtu;
+		PMD_DRV_LOG(DEBUG, "RNDIS MTU is %u", dev->data->mtu);
+	}
+
+	return ret;
+
+error:
+	/* In case of error, attempt to restore original MTU */
+	if (hn_attach(hv, orig_mtu))
+		PMD_DRV_LOG(ERR, "Restoring original MTU failed");
+
```

Comment:
In case of error, need to restore all subchannels on original mtu

Reply:
I've refactored this so all subchannels would be restored if we are attempting to rollback to the original MTU.